### ISSUE TITLE
Slicing after [u8]::first performs unnessessary bound check

### ISSUE THREAD

**Issue by youknowone** [rust-lang/rust#75268]
Playground link: https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=2e2d548aa3a7b89bfb074456d971450d

```rust
pub fn boundcheck(mut lit: &[u8]) -> Option<&[u8]> {
    if lit.first() == Some(&b'_') {
        lit = &lit[1..];
        // lit = unsafe { lit.get_unchecked(1..) };
    }
    Some(lit)
}
```
This slice always exists because of the branch, but it performs bound check.

Real case: https://github.com/RustPython/RustPython/pull/2077

**Comment by Frizi** [comment]
Interestingly, following code seems to be properly optimized
```rust
pub fn boundcheck(mut lit: &[u8]) -> Option<&[u8]> {
    if let Some(&b'_') = lit.first() {
        lit = &lit[1..];
    }
    Some(lit)
}
```

**Comment by rodrimati1992** [comment]
Another way to write that:
```rust
pub fn boundcheck(mut lit: &[u8]) -> Option<&[u8]> {
    if let [b'_', rem @ ..] = lit {
        lit = rem;
    }
    Some(lit)
}
```
Before Rust 1.42.0:
```rust
pub fn boundcheck(mut lit: &[u8]) -> Option<&[u8]> {
    if let Some((&b'_', rem)) = lit.split_first() {
        lit = rem;
    }
    Some(lit)
}
```

**Comment by nikic** [comment]
Based on a local test #74533 fixes this issue as well.

**Comment by lzutao** [comment]
Closed by #74533 .
The new asm generated by nightly (ceedf1d5f 2020-08-08):
```
check::boundcheck:
 mov     rax, rdi
 test    rsi, rsi
 je      .LBB0_1
 mov     rdx, rsi
 cmp     byte, ptr, [rax], 95
 jne     .LBB0_4
 add     rax, 1
 add     rdx, -1
.LBB0_4:
 ret
.LBB0_1:
 xor     edx, edx
 ret
```

**Comment by youknowone** [comment]
Nice to know this will be fixed. Thank you all.
And thanks for the advice @rodrimati1992
